Common/Securities/Option/StrategyMatcher/OptionPositionCollection.cs

```diff
@@ -0,0 +1,489 @@
+        {
+            var rights = _rights.Remove(right.Invert());
+
+            var positions = ImmutableDictionary<Symbol, OptionPosition>.Empty;
+            if (includeUnderlying && HasUnderlying)
+            {
+                positions = positions.Add(Underlying, UnderlyingPosition);
+            }
+
+            var strikes = ImmutableSortedDictionary<decimal, ImmutableHashSet<Symbol>>.Empty;
+            var expirations = ImmutableSortedDictionary<DateTime, ImmutableHashSet<Symbol>>.Empty;
+            foreach (var symbol in rights.SelectMany(kvp => kvp.Value))
+            {
+                var position = _positions[symbol];
+                positions = positions.Add(symbol, position);
+                strikes = strikes.Add(position.Strike, symbol);
+                expirations = expirations.Add(position.Expiration, symbol);
+            }
+
+            return new OptionPositionCollection(positions, rights, strikes, expirations);
+        }
+
+        /// <summary>
+        /// Slices this collection, returning a new collection containing only
+        /// positions matching the specified <paramref name="comparison"/> and <paramref name="strike"/>
+        /// </summary>
+        public OptionPositionCollection Slice(BinaryComparison comparison, decimal strike, bool includeUnderlying = true)
+        {
+            var strikes = comparison.Filter(_strikes, strike);
+            if (strikes.IsEmpty)
+            {
+                return includeUnderlying && HasUnderlying ? Empty.Add(UnderlyingPosition) : Empty;
+            }
+
+            var positions = ImmutableDictionary<Symbol, OptionPosition>.Empty;
+            if (includeUnderlying)
+            {
+                OptionPosition underlyingPosition;
+                if (_positions.TryGetValue(Underlying, out underlyingPosition))
+                {
+                    positions = positions.Add(Underlying, underlyingPosition);
+                }
+            }
+
+            var rights = ImmutableDictionary<OptionRight, ImmutableHashSet<Symbol>>.Empty;
+            var expirations = ImmutableSortedDictionary<DateTime, ImmutableHashSet<Symbol>>.Empty;
+            foreach (var symbol in strikes.SelectMany(kvp => kvp.Value))
+            {
+                var position = _positions[symbol];
+                positions = positions.Add(symbol, position);
+                rights = rights.Add(symbol.ID.OptionRight, symbol);
+                expirations = expirations.Add(symbol.ID.Date, symbol);
+            }
+
+            return new OptionPositionCollection(positions, rights, strikes, expirations);
+        }
+
+        /// <summary>
+        /// Slices this collection, returning a new collection containing only
+        /// positions matching the specified <paramref name="comparison"/> and <paramref name="expiration"/>
+        /// </summary>
+        public OptionPositionCollection Slice(BinaryComparison comparison, DateTime expiration, bool includeUnderlying = true)
```

Comment:
I'm not sure I follow -- maybe a quick 'n dirty code sample?

Reply:
Sorry :D I edited your post -- when posting code blocks, you can specify the language. You do the normal three backticks and then can add 'csharp' like:

``` text
//   ``` csharp
//   ```
```

This gives you C# specific syntax highlighting

Reply:
This would work and is similar to the `ForSymbols` type methods (`ForRight`, `ForSide`, `ForStrike`, `ForExpiration`). The main reason I didn't go this route is due to the extra work. Take `Slice(OptionRight)` for example, by accessing `_rights[right]` we already have an `ImmutableHashSet<Symbol>` which can be directly used to create the new `ImmutableDictionary<OptionRight, ImmutableHashSet<Symbol>` for the new collection instance. If we refactor/extract as proposed, we would needlessly be recreating that hashset. Since these `Slice` methods are expected to be the hottest parts of this implementation, I made an effort to not recreate collections unless necessary. This does lead to methods that look very similar and could be refactored and a few ways. For now let's leave it as-is -- it may turn out that performance here is negligible and the hottest blocks are elsewhere, in which case we can come back in here and refactor for readability/maintainability.

Reply:
> Since these Slice methods are expected to be the hottest parts of this implementation, I made an effort to not recreate collections unless necessary.

hm each `Add` is already creating a new collection 😛 

> For now let's leave it as-is -- it may turn out that performance here is negligible and the hottest blocks are elsewhere, in which case we can come back in here and refactor for readability/maintainability.

Yeah sounds good 👍